src/boot/memory.c

```diff
@@ -115,139 +110,147 @@ void move_segment_table_to_dmem(void) {
 }
 #endif
 
+extern u8 __mainPoolStart[];
+
 /**
- * Initialize the main memory pool. This pool is conceptually a pair of stacks
+ * Initialize the main memory pool. This pool is conceptually regions
  * that grow inward from the left and right. It therefore only supports
  * freeing the object that was most recently allocated from a side.
  */
-void main_pool_init(void *start, void *end) {
-    sPoolStart = (u8 *) ALIGN16((uintptr_t) start) + 16;
-    sPoolEnd = (u8 *) ALIGN16((uintptr_t) end - 15) - 16;
-    sPoolFreeSpace = sPoolEnd - sPoolStart;
-
-    sPoolListHeadL = (struct MainPoolBlock *) (sPoolStart - 16);
-    sPoolListHeadR = (struct MainPoolBlock *) sPoolEnd;
-    sPoolListHeadL->prev = NULL;
-    sPoolListHeadL->next = NULL;
-    sPoolListHeadR->prev = NULL;
-    sPoolListHeadR->next = NULL;
+void main_pool_init() {
+#define SET_REGION(id, bufStart, bufEnd) \
+    gMainPool.regions[id].start = (u8 *) ALIGN4((uintptr_t)(bufStart)); \
+    gMainPool.regions[id].end = (u8 *) DOWN4((uintptr_t)(bufEnd));
+
+    SET_REGION(0, __mainPoolStart, RAM_END);
+
+#undef SET_REGION
```

Comment:
This seems needlessly messy. Even in the cases of multiple memory regions, there's gotta be a better way...

Reply:
I have reworked it to just use a simple function `main_pool_region_init`. I believe it is more readable now.